C_hands_on_programming/ipc_methods.c

```diff
@@ -0,0 +1,358 @@
+				printf("<message is used>\n");
+				info.method = message;
+			} else if(pflag){
+				printf("<pipe is used>\n");
+				info.method = pipes;
+			} else if (qflag) {
+				printf("<queue is used>\n");
+				info.method = queue;
+			} else {
+				printf("<shm is used>\n");
+				info.method = shm;
+			}
+		}
+	} else {
+		printf("A filename must be given as an argument for option --file/-f(only one -f is allowed)\n");
+		exit(EXIT_FAILURE);
+	}
+	return info;
+}
+// This version of messageSend implemented using read()
+#if 0
+void messageSend(const ipc_info *info) {
+	printf("Starting messageSend..\n"
+			"If client specified server name does not exist, this program is stuck in an infinite loop looking for a server\n"
+			);
+
+	int fd, current, coid, status, size;
+    header_t header;
+    char* buffer = NULL;
+    iov_t siov[2];
+	// open file for reading
+	fd = open(info->filename, O_RDONLY);
+	if (fd == -1) {
+		perror("open\n");
+		exit(EXIT_FAILURE);
+	}
+
+	printf("Looking for server..\n");
+    //Server search loop
+    while ((coid = name_open(info->argument_string, 0)) == -1)
+    {
+        sleep(1);
+    }
+    printf("Found the server connection!\n");
+
+    //Start message reading
+
+    current = lseek(fd, 0, SEEK_CUR);
+    size = lseek(fd, 0, SEEK_END);
+    //set the file pointer back to beginning of file
+    lseek(fd, current, SEEK_SET);
+
+	header.msg_type = HEADER_IOV_TYPE;
+	header.data_size = size;
+
+    if ((buffer = (char*)malloc(size)) == NULL) {
+    	perror("malloc\n");
+    	free(buffer);
+    	exit(EXIT_FAILURE);
+    }
+
+	//read the file
+    if (read(fd, buffer, size) == -1) {
+    	perror("read\n");
+    	exit(EXIT_FAILURE);
+    }
+
+    close(fd);
+
+	SETIOV(&siov[0], &header, sizeof(header));
+	SETIOV(&siov[1], buffer, header.data_size);
+
+	status = MsgSendvs(coid, siov, 2, NULL, 0);
+	if (status == -1) {
+		perror("MsgSendvs");
+		exit(EXIT_FAILURE);
+	}
+	free(buffer);
+
+	if (MsgSendPulse(coid, -1, EOF_PULSE_CODE, 0) == -1) {
+		perror("MsgSendPulse\n");
+		exit(EXIT_FAILURE);
+	} else {
+		printf("Exiting the program..\n");
+	}
+
+}
+#endif
+
+//This version of messageSend implemented using fread()
+#if 1
+void messageSend(const ipc_info *info) {
+	printf("Starting messageSend..\n"
+			"If client specified server name does not exist, this program is stuck in an infinite loop looking for a server\n"
+			);
+
+    int coid, status, size;
+    header_t header;
+    char* buffer = NULL;
+    iov_t siov[2];
+	FILE *read;
+	// open file for reading
+	read = fopen(info->filename, "rb");
+	if (read == NULL) {
+		printf("File name does not exist! Exiting the program..\n");
+		exit(EXIT_FAILURE);
+	}
+
+	printf("Looking for server..\n");
+    //Server search loop
+    while ((coid = name_open(info->argument_string, 0)) == -1)
+    {
+        sleep(1);
+    }
+    printf("Found the server connection!\n");
+
+    //Start message reading
+	fseek(read, 0, SEEK_END);
+	size = ftell(read);
+	fseek(read, 0, SEEK_SET);
```

Comment:
fseek should be avoid to find the size of a file. You can use instead ```struct stat()```.

Reply:
Why is fseek not recommended?

Reply:
https://wiki.sei.cmu.edu/confluence/display/c/FIO19-C.+Do+not+use+fseek()+and+ftell()+to+compute+the+size+of+a+regular+file 
Farnood have sent this link in the team chat/